File: leetcode_C++/p1921.cpp
/*
@Author: mrxhar
@Date:   2023-09-03 08:58:46
@Email:  1337904145@qq.com
@Blog:   https://leetcode.cn/problems/eliminate-maximum-number-of-monsters/?envType=daily-question&envId=2023-09-03
*/
#include <vector>
using std::vector;
#include <algorithm>

class Solution
{
public:
    int eliminateMaximum(vector<int> &dist, vector<int> &speed)
    {
        int n = dist.size();
        vector<int> arrivalTimes(n);
        for (int i = 0; i < n; i++)
        {
            arrivalTimes[i] = (dist[i] - 1) / speed[i] + 1;
        }
        sort(arrivalTimes.begin(), arrivalTimes.end());
        for (int i = 0; i < n; i++)
        {
            if (arrivalTimes[i] <= i)
            {
                return i;
            }
        }
        return n;
    }
};

File: leetcode_py/p1921.py
from typing import List


class Solution:
    def eliminateMaximum(self, dist: List[int], speed: List[int]) -> int:
        n = len(dist)
        arrTime = [0] * n
        for i in range(n):
            arrTime[i] = (dist[i] - 1) // speed[i] + 1
        arrTime.sort()
        for i in range(n):
            if arrTime[i] < i:
                return i
        return n
